mesonbuild/dependencies/mpi.py

```diff
@@ -37,65 +38,90 @@ def mpi_factory(env: 'Environment',
-            # fall back to the hardcoded names
-            if language == 'c':
-                tool_names = [os.environ.get('MPICC'), 'mpicc']
+                tool_names.append('mpiicc')
             elif language == 'cpp':
-                tool_names = [os.environ.get('MPICXX'), 'mpic++', 'mpicxx', 'mpiCC']
+                tool_names.append('mpiicpc')
             elif language == 'fortran':
-                tool_names = [os.environ.get(e) for e in ['MPIFC', 'MPIF90', 'MPIF77']]
-                tool_names.extend(['mpifort', 'mpif90', 'mpif77'])
+                tool_names.append('mpiifort')
 
-            cls = OpenMPIConfigToolDependency
-
-        tool_names = [t for t in tool_names if t]  # remove empty environment variables
-        assert tool_names
+        # even with intel compilers, mpicc has to be considered
+        if language == 'c':
+            tool_names.append('mpicc')
+        elif language == 'cpp':
+            tool_names.extend(['mpic++', 'mpicxx', 'mpiCC'])
+        elif language == 'fortran':
+            tool_names.extend(['mpifort', 'mpif90', 'mpif77'])
 
         nwargs['tools'] = tool_names
         candidates.append(functools.partial(
-            cls, tool_names[0], env, nwargs, language=language))
+            MPIConfigToolDependency, tool_names[0], env, nwargs, language=language))
 
-    if DependencyMethods.SYSTEM in methods:
+    if DependencyMethods.SYSTEM in methods and env.machines[for_machine].is_windows():
         candidates.append(functools.partial(
             MSMPIDependency, 'msmpi', env, kwargs, language=language))
 
+    # Only OpenMPI has pkg-config, and it doesn't work with the intel compilers
+    # for MPI, environment variables and commands like mpicc should have priority
+    if DependencyMethods.PKGCONFIG in methods and not compiler_is_intel:
+        pkg_name = None
+        if language == 'c':
+            pkg_name = 'ompi-c'
+        elif language == 'cpp':
+            pkg_name = 'ompi-cxx'
+        elif language == 'fortran':
+            pkg_name = 'ompi-fort'
+        candidates.append(functools.partial(
+            PkgConfigDependency, pkg_name, env, kwargs, language=language))
+
     return candidates
 
 packages['mpi'] = mpi_factory
 
 
-class _MPIConfigToolDependency(ConfigToolDependency):
+class MPIConfigToolDependency(ConfigToolDependency):
+    """Wrapper around mpicc, Intel's mpiicc and friends."""
+
+    def __init__(self, name: str, env: 'Environment', kwargs: T.Dict[str, T.Any],
+                 language: T.Optional[str] = None):
+        super().__init__(name, env, kwargs, language=language)
+        if not self.is_found:
+            return
+
+        for comp, link in [('--showme:compile', '--showme:link'), ('-compile_info', '-link_info'), ('-show', None)]:
+            try:
+                c_args = self.get_config_value([comp], 'compile_args')
+                l_args = self.get_config_value([link], 'link_args') if link is not None else c_args
```

Comment:
Can you document (in a comment) for future readers, which compilers accept which forms of the arguments here?

Reply:
That would be e.g. mpich for -compile_info per https://github.com/mesonbuild/meson/issues/13615#issuecomment-2320541132

Reply:
I added a comment.